{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1 style = \"text-align: center; color: green; font-size: 32px\"> <b> PreProt üß¨ </b> </h1>\n",
    "<p style = \"text-align: center; color: cyan; font-size: 18px\"> Modelo de Machine Learning para predi√ß√£o de prote√≠nas da <i> E. Coli </i> </p>\n",
    "<p style = \"text-align: center; font-size: 12px\"> Projeto de conclus√£o para o curso de Data Science & analytics pela USP/ESALQ </p>\n",
    "<p style = \"text-align: center; font-size: 10px\"> Desenvolvido por Fernando Falat, orientado por Miriam Martin </p>\n",
    "<b style = \"font-size: 24px\" > O que √© o PreProt? </b>\n",
    "<p style = \"font-size: 14px\"> PreProt √© um projeto de implementa√ß√£o de Machine Learning, mais especificamente Naive Bayes, para avaliar a efic√°cia e acur√°cia do modelo dentro do campo da bioinform√°tica, atrav√©s de testes de efic√°cia como o F test e outros par√¢metros. Mais informa√ß√µes do projeto podem ser encontradas na <a href = \"readme.txt\"> documenta√ß√£o </a> e na <a href = \"https://docs.google.com/document/d/1evF9-wIk1tZ6xFhq2hgq4nII7toi9ARP/edit)\"> monografia </a>. </p>"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3 style = \"font-size: 24px\"> <b> Formato de arquivos </b> </h3>\n",
    "<p style = \"font-size: 14px\"> Na bioinform√°tica, sequ√™ncias de DNA e prote√≠nas s√£o os tipos de dados mais comuns e por padr√£o s√£o armazenadas em um formato de arquivo \"FASTA\" (fast-all), \"FAST-P\" (prote√≠nas) ou \"FAST-N\" (nucleot√≠deos).\n",
    "<br></br>\n",
    "<b> DNA: </b> A,T,C,G\n",
    "<br></br>\n",
    "<b> AMINO√ÅCIDOS: </b>Acr√¥nimos de letras, ex: A para alanina </p>\n",
    "<p style = \"font-size: 14px\"> Para abrir arquivos nesse formato, utiliza-se o SeqIO da biblioteca BioPython. Essa classe possibilta uma interface para trabalhar com esses tipos de dados. </p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Importando biblioteca para leitura de arquivos FASTA\n",
    "from Bio import SeqIO\n",
    "\n",
    "# Abrindo o arquivo e apresentando algumas informa√ß√µes\n",
    "sequences = [] # Criando uma lista vazia\n",
    "for seq_record in SeqIO.parse(\"datasets/protein_sequences.fasta\", \"fasta\"):\n",
    "    # Adicionando o record na lista vazia\n",
    "    sequences.append(str(seq_record.seq))\n",
    "    # printando a sequ√™ncia\n",
    "    print(seq_record.seq)\n",
    "    # printando o identificador da sequ√™ncia (prote√≠na)\n",
    "    print(seq_record.id)\n",
    "    # printando o tamanho da sequ√™ncia (cadeia de amino√°cidos)\n",
    "    print(len(seq_record))\n",
    "print(\"N√∫mero de sequ√™ncias: \", len(sequences))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sequencia de amino√°cidos da E. Coli\n",
    "print(sequences)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<style>\n",
    "  p {\n",
    "    text-align: justify;\n",
    "    font-size: 14px !important;\n",
    "  }\n",
    "</style>\n",
    "\n",
    "<h3 style = \"font-size: 24px\"> <b> Introdu√ß√£o </b> </h3>\n",
    "\n",
    "<p style = \"font-size: 14px\"> O output fornecido acima representa a sequ√™ncia de amino√°cidos presentes no genoma da cepa de <i>E. Coli</i> em quest√£o. Podemos ver que cada sequ√™ncia de amino√°cidos representa uma prote√≠na com nome \"ecmdb_XXXXXX\".</p>\n",
    "<div style = \"font-size: 14px\">\n",
    "<p align = \"justify\" style = \"font-size: 14px;\"> A fim de que ocorra a forma√ß√£o de uma prote√≠na, √© necess√°rio percorrer um conjunto de etapas. Iniciando-se pelos √°cidos nucleicos, que constituem a base do DNA (ATCG):\n",
    "<ol type = \"1\">\n",
    "<li> <b> Transcri√ß√£o: </b> O DNA √© transicro em RNA mensageiro (mRNA) atrav√©s da RNA polimerase.</li> \n",
    "<br></br>\n",
    "<li> <b> Tradu√ß√£o: </b> O mRNA √© lido pelos ribossomos, decodificando os c√≥dons (sequ√™ncia de 3 √°cidos nucleicos), ent√£o entram os tRNAs e transportam os amino√°cidos correspondentes para formar a cadeia polipept√≠dica. Um pept√≠deo √© composto por dois ou mais amino√°ciodos, sendo classificados em dipept√≠deos (2); tri; tetra; oligo e polipept√≠deos; quando a cadeia de amino√°cidos passa de 70, classifica-se em prote√≠na.</li> \n",
    "<br></br>\n",
    "<li> <b>Desdobramento e p√≥s-tradu√ß√£o: </b> Ap√≥s a s√≠ntese da cadeia polipept√≠dica, a prote√≠na pode passar por um processo de dobramento ou enovelamento para adquirir sua estrutura tridimensional funcional. Ap√≥s o dobramento, a prote√≠na pode sofrer modifica√ß√µes p√≥s-traducionais, como a adi√ß√£o de grupos qu√≠micos ou a clivagem de segmentos adicionais. </li>\n",
    "</ol>\n",
    "</p>\n",
    "<br></br>\n",
    "<p style = \"font-size: 14px\"> A imagem abaixo exemplifica o processo da tradu√ß√£o da fita de mRNA at√© a leitura pela fita de tRNA: </p>\n",
    "<p style=\"text-align:center;\"> <img src = \"https://upload.wikimedia.org/wikipedia/commons/thumb/7/70/Aminoacids_table.svg/609px-Aminoacids_table.svg.png?20210405175054\" alt = \"Tabela de convers√£o de amino√°cidos\" width = \"500\" heigh = \"500\" align = center> </img> </p>\n",
    "</div>\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "''' Fun√ß√µes de convers√£o '''\n",
    "\n",
    "# Fun√ß√£o de DNA para mRNA\n",
    "def dna_para_mRNA(sequencia_DNA):\n",
    "\n",
    "    # Transcreve a sequ√™ncia de DNA para mRNA atrav√©s da substitui√ß√£o do T pelo U utilizando a fun√ß√£o replace.\n",
    "\n",
    "    return sequencia_DNA.replace('T', 'U')\n",
    "\n",
    "# Fun√ß√£o de mRNA para tRNA\n",
    "def mRNA_para_tRNA(sequencia_mRNA):\n",
    "\n",
    "    # Transcreve o mRNA para tRNA atrav√©s da troca de cada nucleot√≠deo com seu respectivo par pelas regras de pareamento c√≥don-antic√≥don\n",
    "    \n",
    "    dicionario_aux = {'A': 'U', 'U': 'A', 'C': 'G', 'G': 'C'}\n",
    "    sequencia_tRNA = \"\"\n",
    "    \n",
    "    #Loop para itera√ß√£o na vari√°vel\n",
    "    for nucleotideo in sequencia_mRNA:\n",
    "        if nucleotideo in dicionario_aux:\n",
    "            sequencia_tRNA += dicionario_aux[nucleotideo]\n",
    "        else:\n",
    "            sequencia_tRNA += nucleotideo\n",
    "    \n",
    "    return sequencia_tRNA\n",
    "\n",
    "# Fun√ß√£o de mRNA e tRNA para amino√°cidos\n",
    "def traducao(sequencia_mRNA):\n",
    "\n",
    "    # Traduz uma sequ√™ncia de mRNA para a respectiva sequ√™ncia de amino√°cidos atrav√©s do mapeamento da trinca de nucleot√≠deos (c√≥dons)  \n",
    "\n",
    "    tabela_codon = {\n",
    "        'AUG': 'M', 'UUU': 'F', 'UUC': 'F', 'UUA': 'L', 'UUG': 'L',  # ... Adicionar mais c√≥dons e seus respectivos amino√°cidos\n",
    "        'UCU': 'S', 'UCC': 'S', 'UCA': 'S', 'UCG': 'S', 'UAU': 'Y',\n",
    "        'UAC': 'Y', 'UAA': '*', 'UAG': '*', 'UGU': 'C', 'UGC': 'C',\n",
    "        'UGA': '*', 'UGG': 'W', 'CUU': 'L', 'CUC': 'L', 'CUA': 'L',\n",
    "        'CUG': 'L', 'CCU': 'P', 'CCC': 'P', 'CCA': 'P', 'CCG': 'P',\n",
    "        'CAU': 'H', 'CAC': 'H', 'CAA': 'Q', 'CAG': 'Q', 'CGU': 'R',\n",
    "        'CGC': 'R', 'CGA': 'R', 'CGG': 'R', 'AUU': 'I', 'AUC': 'I',\n",
    "        'AUA': 'I', 'AUC': 'I', 'ACU': 'T', 'ACC': 'T', 'ACA': 'T',\n",
    "        'ACG': 'T', 'AAU': 'N', 'AAC': 'N', 'AAA': 'K', 'AAG': 'K',\n",
    "        'AGU': 'S', 'AGC': 'S', 'AGA': 'R', 'AGG': 'R', 'GUU': 'V',\n",
    "        'GUC': 'V', 'GUA': 'V', 'GUG': 'V', 'GCU': 'A', 'GCC': 'A',\n",
    "        'GCA': 'A', 'GCG': 'A', 'GAU': 'D', 'GAC': 'D', 'GAA': 'E',\n",
    "        'GAG': 'E', 'GGU': 'G', 'GGC': 'G', 'GGA': 'G', 'GGG': 'G'\n",
    "    }\n",
    "    \n",
    "    sequencia_proteina = \"\"\n",
    "    \n",
    "    # Loop de itera√ß√£o da lista mRNA de input com a tabela de codons, for√ßando a leitura por codons\n",
    "    for i in range(0, len(sequencia_mRNA), 3):\n",
    "        codon = sequencia_mRNA[i:i+3]\n",
    "        \n",
    "        if codon in tabela_codon:\n",
    "            amino_acido = tabela_codon[codon]\n",
    "            if amino_acido == '*':  # Stop Codon encontrado\n",
    "                break\n",
    "            sequencia_proteina += amino_acido\n",
    "    \n",
    "    return sequencia_proteina\n",
    "\n",
    "# confirma√ß√£o que as fun√ß√µes foram declaradas\n",
    "print(\"Fun√ß√µes OK.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Simuala√ß√£o de uma replica√ß√£o\n",
    "\n",
    "sequencia_DNA = \"ATGCATCGTAA\"\n",
    "sequencia_mRNA = dna_para_mRNA(sequencia_DNA)\n",
    "sequencia_tRNA = mRNA_para_tRNA(sequencia_mRNA)\n",
    "sequencia_proteina = traducao(sequencia_mRNA) # sequ√™ncia de prote√≠nas >= 70 Amino√°cidos\n",
    "\n",
    "print(\"Sequ√™ncia de DNA:\", sequencia_DNA)\n",
    "print(\"Sequ√™ncia de mRNA:\", sequencia_mRNA)\n",
    "print(\"Sequ√™ncia de tRNA:\", sequencia_tRNA)\n",
    "print(\"Sequ√™ncia de amino√°cidos:\", sequencia_proteina) "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<style>\n",
    "  p {\n",
    "    text-align: justify !important;\n",
    "    font-size: 14px !important;\n",
    "  }\n",
    "\n",
    "  h3 {\n",
    "  font-size: 24px\n",
    "  }\n",
    "</style>\n",
    "\n",
    "<h3> <b> Machine Learning </b> </h3>\n",
    "\n",
    "<div style = \"font-size: 14px\">\n",
    "<p> De acordo com a literatura cient√≠fica, uma prote√≠na √© composta por uma cadeia polipept√≠dica que pode conter mais de 70 amino√°cidos. Como existem 20 tipos diferentes de amino√°cidos que podem ser usados para construir uma prote√≠na, e esses amino√°cidos podem ser combinados em qualquer ordem, o n√∫mero de sequ√™ncias poss√≠veis √© virtualmente infinito. Portanto, as prote√≠nas possuem combina√ß√µes quase infinitas de amino√°cidos. </p>\n",
    "\n",
    "<p> Logo mm algoritmo de machine learning pode aprender a partir de dados experimentais como o DNA √© transcrito em RNA e depois traduzido em prote√≠nas, quais s√£o os fatores que influenciam essa transforma√ß√£o e quais s√£o as consequ√™ncias de muta√ß√µes ou altera√ß√µes na express√£o g√™nica.</p>\n",
    "\n",
    "<p> Com o objetivo de facilitar o processo de tradu√ß√£o de uma sequ√™ncia de DNA em uma prote√≠na, o PreProt treina um modelo de aprendizado de m√°quina, e analisa os resultados, com uma base de dados contendo o proteoma da <i>Escherichia coli</i> K-12 substr. MG1655, que √© a bact√©ria mais estudada na humanidade. </p>\n",
    "\n",
    "</div>\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "''' Carregamento e Tratamento dos dados'''\n",
    "\n",
    "# Importando e tratando o dataset contendo o proteoma da E. Coli K-12 MG1655 em formato FASTA\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "\n",
    "# Fun√ß√£o para ler o arquivo FASTA e organizar os dados para criar um dataframe\n",
    "def ler_fasta(file_path):\n",
    "    sequences = [] # Criando uma lista vazia\n",
    "    for record in SeqIO.parse(file_path, \"fasta\"): # iterando a base de dados na lista vazia\n",
    "        sequences.append({\"Sequencia_AA\": str(record.seq), \"ID_Proteina\": record.id}) # criando um vetor bi-dimensional \n",
    "\n",
    "    return sequences\n",
    "\n",
    "# Inputs\n",
    "file_path = \"datasets/Escherichia coli str. K-12 substr. MG1655/proteome.fasta\"\n",
    "sequencias = ler_fasta(file_path)\n",
    "\n",
    "# Criando um Pandas Dataframe a partir do vetor sequences para posterior treinamento do modelo\n",
    "df = pd.DataFrame(sequencias)\n",
    "# Removendo a parte n√£o utilizada do ID da prote√≠na na coluna\n",
    "df['ID_Proteina'] = df['ID_Proteina'].str.rsplit('|', n=1).str[-1]\n",
    "\n",
    "print(df.head())\n",
    "print(len(df.index)) #6463 prote√≠nas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Analisando o proteoma da E. Coli K-12\n",
    "\n",
    "# Importando biblioteca para leitura de arquivos FASTA\n",
    "from Bio import SeqIO\n",
    "\n",
    "# Abrindo o arquivo e apresentando algumas informa√ß√µes\n",
    "sequences = [] # Criando uma lista vazia\n",
    "for seq_record in SeqIO.parse(\"datasets/Escherichia coli str. K-12 substr. MG1655/proteome.fasta\", \"fasta\"):\n",
    "    # Adicionando o record na lista vazia\n",
    "    sequences.append(str(seq_record.seq))\n",
    "    # printando a sequ√™ncia\n",
    "    print(seq_record.seq)\n",
    "    # printando o identificador da sequ√™ncia (prote√≠na)\n",
    "    print(seq_record.id)\n",
    "    # printando o tamanho da sequ√™ncia (cadeia de amino√°cidos)\n",
    "    print(len(seq_record))\n",
    "\n",
    "print(\"Sequ√™ncias no arquivo:\", len(sequences))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# C√≥digo para o modelo de aprendizado de maquina - Classificador de Naive Bayes\n",
    "# Escolha do modelo por ser um classificador de categorias\n",
    "\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.feature_extraction.text import CountVectorizer\n",
    "from sklearn.naive_bayes import MultinomialNB\n",
    "\n",
    "# Dividindo o Dataset em subsets de treino e teste\n",
    "X_train, X_test, y_train, y_test = train_test_split(df[\"Sequencia_AA\"], df[\"ID_Proteina\"], test_size=0.2, random_state=42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PRIMEIRA VERS√ÉO\n",
    "# Extra√ß√£o de caracter√≠sticas \n",
    "vectorizer = CountVectorizer()\n",
    "X_train_encoded = vectorizer.fit_transform(X_train)\n",
    "X_test_encoded = vectorizer.transform(X_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SEGUNDA VERS√ÉO\n",
    "# Trocando o Encoding para melhorar acur√°cia: k-mer length\n",
    "from sklearn.feature_extraction.text import CountVectorizer\n",
    "\n",
    "# Step 2: Convert Protein Sequences to Numerical Features (k-mer frequency encoding)\n",
    "k = 1  # Specify the length of k-mers\n",
    "\n",
    "# Convert protein sequences to k-mer frequency representation\n",
    "vectorizer = CountVectorizer(analyzer='char', ngram_range=(k, k))\n",
    "X_train_encoded = vectorizer.fit_transform(X_train)\n",
    "X_test_encoded = vectorizer.transform(X_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TERCEIRA VERS√ÉO\n",
    "# Multiple sequence alignment (MSA) - O m√©todo utilizado para o caso de sequ√™ncia de prote√≠nas em machine learning\n",
    "# primeiramente realizando um PSI-BLAST https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastSearch&PROGRAM=blastp&BLAST_PROGRAMS=psiBlast com o proteoma como input\n",
    "# O PSI-BLAST tem como objetivo encontrar prote√≠nas com sequ√™ncias e fun√ß√µes parecidas, gerando uma position-specific scoring matrices (PSSMs)\n",
    "# Ent√£o essa PSSMs pode ser utilizada para treinar nosso algor√≠timo de classifica√ß√£o\n",
    "# Para iterar o PSI-BLAST para cada prote√≠na do proteoma da E. Coli K-12 necessitaria de MUITO poder computacional, logo vou baixar uma vers√£o de PSSM do NCBI\n",
    "# O uso desse modelo nos permite chegar a sequencia, estrutura e fun√ß√£o da prote√≠na!! https://www.ncbi.nlm.nih.gov/Structure/cdd/cdd.shtml\n",
    "# Input do proteoma aqui https://www.ncbi.nlm.nih.gov/Structure/cdd/cdd_help.shtml#BatchRPSBInput\n",
    "# Checar arquivo pssm.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Treinando o modelo\n",
    "naive_bayes_model = MultinomialNB()\n",
    "naive_bayes_model.fit(X_train_encoded, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Aplicando m√©tricas de an√°lise do modelo\n",
    "from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score\n",
    "\n",
    "y_pred_train = naive_bayes_model.predict(X_train_encoded)\n",
    "train_accuracy = accuracy_score(y_train, y_pred_train)\n",
    "train_precision = precision_score(y_train, y_pred_train, average='weighted')\n",
    "train_recall = recall_score(y_train, y_pred_train, average='weighted')\n",
    "train_f1 = f1_score(y_train, y_pred_train, average='weighted')\n",
    "\n",
    "print(\"Acur√°cia:\", train_accuracy)\n",
    "print(\"Precis√£o:\", train_precision)\n",
    "print(\"Recall:\", train_recall)\n",
    "print(\"F1-Score:\", train_f1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Criando gr√°ficos e visualiza√ß√µes do modelo\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "metricas = ['Acur√°cia', 'Precis√£o', 'Recall', 'F1-Score']\n",
    "valores = [train_accuracy, train_precision, train_recall, train_f1]\n",
    "valores_percentuais = [value * 100 for value in values]\n",
    "\n",
    "plt.figure(figsize=(8, 4))\n",
    "plt.bar(metricas, valores_percentuais, color='steelblue')\n",
    "plt.xlabel('M√©tricas')\n",
    "plt.ylabel('Valores')\n",
    "plt.title('Avalia√ß√£o do Modelo - M√©tricas')\n",
    "plt.ylim([0, 1])\n",
    "\n",
    "# Set y-axis tick positions and labels as percentages\n",
    "plt.yticks(np.linspace(0, 100, 11), ['{}%'.format(int(valores_percentuais)) for valores_percentuais in np.linspace(0, 100, 11)])\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Predi√ß√µes\n",
    "nova_sequencia = [\"MFENITAAPADPILGLADLFRADERPGKINLGIGVYKDETGKTPVLTSVKKAEQYLLENETTKNYLGIDGIPEFGRCTQELLFGKGSALINDKRARTAQTPGGTGALRVAADFLAKNTSVKRVWVSNPSWPNHKSVFNSAGLEVREYAYYDAENHTLDFDALINSLNEAQAGDVVLFHGCCHNPTGIDPTLEQWQTLAQLSVEKGWLPLFDFAYQGFARGLEEDAEGLRAFAAMHKELIVASSYSKNFGLYNERVGACTLVAADSETVDRAFSQMKAAIRANYSNPPAHGASVVATILSNDALRAIWEQELTDMRQRIQRMRQLFVNTLQEKGANRDFSFIIKQNGMFSFSGLTKEQVLRLREEFGVYAVSGRVNVAGMTPDNMAPLCEAIVAVL\"] # NOVO INPUT\n",
    "# essa √© a sequ√™ncia de amino√°cidos da prote√≠na I\n",
    "\n",
    "# INSA9_ECOLI\n",
    "# MASVSISCPSCSATDGVVRNGKSTAGHQRYLCSHCRKTWQLQFTYTASQPGTHQKIIDMAMNGVGCRATARIMGVGLNTILRHLKNSGRSR\n",
    "\n",
    "# AAT_ECOLI\n",
    "# MFENITAAPADPILGLADLFRADERPGKINLGIGVYKDETGKTPVLTSVKKAEQYLLENETTKNYLGIDGIPEFGRCTQELLFGKGSALINDKRARTAQTPGGTGALRVAADFLAKNTSVKRVWVSNPSWPNHKSVFNSAGLEVREYAYYDAENHTLDFDALINSLNEAQAGDVVLFHGCCHNPTGIDPTLEQWQTLAQLSVEKGWLPLFDFAYQGFARGLEEDAEGLRAFAAMHKELIVASSYSKNFGLYNERVGACTLVAADSETVDRAFSQMKAAIRANYSNPPAHGASVVATILSNDALRAIWEQELTDMRQRIQRMRQLFVNTLQEKGANRDFSFIIKQNGMFSFSGLTKEQVLRLREEFGVYAVSGRVNVAGMTPDNMAPLCEAIVAVL\n",
    "\n",
    "# Se for uma sequ√™ncia de DNA, usar as fun√ß√µes de tradu√ß√£o: fazer um if no algor√≠timo\n",
    "\n",
    "sequencia_mRNA = dna_para_mRNA(sequencia_DNA)\n",
    "sequencia_tRNA = mRNA_para_tRNA(sequencia_mRNA)\n",
    "sequencia_proteina = traducao(sequencia_mRNA)\n",
    "\n",
    "nova_sequencia_encoded = vectorizer.transform(nova_sequencia)\n",
    "nome_proteina = naive_bayes_model.predict(nova_sequencia_encoded)\n",
    "print(\"Nome da poss√≠vel prote√≠na:\", nome_proteina)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession_STEPdb_2.0</th>\n",
       "      <th>Protein_ID_STEPdb_2.0</th>\n",
       "      <th>1ary_Gene_Name</th>\n",
       "      <th>STEPdb_Universal_Name</th>\n",
       "      <th>Uniprot_Protein_Name</th>\n",
       "      <th>STEPdb_Sub-cellular_Location_Full Name</th>\n",
       "      <th>STEPdb_Sub-cellular_Location_Letter Code</th>\n",
       "      <th>sub-cellular_topology_group</th>\n",
       "      <th>SRP_interactome_Schibich2016</th>\n",
       "      <th>TF_interactome_Oh2011</th>\n",
       "      <th>...</th>\n",
       "      <th>TF_client_Martinez2009</th>\n",
       "      <th>TF_client_Butland2005</th>\n",
       "      <th>TF_client_Arifuzzaman2006</th>\n",
       "      <th>TF_client_Niwa2009</th>\n",
       "      <th>GroEL_client_Kerner2005</th>\n",
       "      <th>GroEL_client_Arifuzzaman2006</th>\n",
       "      <th>GroEL_client_Chapman2006</th>\n",
       "      <th>GroEL_client_Niwa2009</th>\n",
       "      <th>SecB_client_Stepdb1.0</th>\n",
       "      <th>SecB_client_additional_data</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P0DPC5</td>\n",
       "      <td>YTID_ECOLI</td>\n",
       "      <td>ytiD</td>\n",
       "      <td>cytoplasmic protein, YtiD</td>\n",
       "      <td>Protein YtiD</td>\n",
       "      <td>Cytoplasmic</td>\n",
       "      <td>A</td>\n",
       "      <td>Cytoplasmic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P0DPC4</td>\n",
       "      <td>YTIC_ECOLI</td>\n",
       "      <td>ytiC</td>\n",
       "      <td>cytoplasmic protein, YtiC</td>\n",
       "      <td>Protein YtiC</td>\n",
       "      <td>Cytoplasmic</td>\n",
       "      <td>A</td>\n",
       "      <td>Cytoplasmic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P0DN74</td>\n",
       "      <td>YTIA_ECOLI</td>\n",
       "      <td>ytiA</td>\n",
       "      <td>cytoplasmic protein, YtiA</td>\n",
       "      <td>Uncharacterized protein YtiA</td>\n",
       "      <td>Cytoplasmic</td>\n",
       "      <td>A</td>\n",
       "      <td>Cytoplasmic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>V9HVX0</td>\n",
       "      <td>YPAA_ECOLI</td>\n",
       "      <td>ypaA</td>\n",
       "      <td>cytoplasmic protein, YpaA</td>\n",
       "      <td>Uncharacterized protein YpaA</td>\n",
       "      <td>Cytoplasmic</td>\n",
       "      <td>A</td>\n",
       "      <td>Cytoplasmic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P0DPC9</td>\n",
       "      <td>YNFQ_ECOLI</td>\n",
       "      <td>ynfQ</td>\n",
       "      <td>cytoplasmic protein, YnfQ</td>\n",
       "      <td>Protein YnfQ</td>\n",
       "      <td>Cytoplasmic</td>\n",
       "      <td>A</td>\n",
       "      <td>Cytoplasmic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows √ó 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Accession_STEPdb_2.0 Protein_ID_STEPdb_2.0 1ary_Gene_Name   \n",
       "0               P0DPC5            YTID_ECOLI           ytiD  \\\n",
       "1               P0DPC4            YTIC_ECOLI           ytiC   \n",
       "2               P0DN74            YTIA_ECOLI           ytiA   \n",
       "3               V9HVX0            YPAA_ECOLI           ypaA   \n",
       "4               P0DPC9            YNFQ_ECOLI           ynfQ   \n",
       "\n",
       "       STEPdb_Universal_Name          Uniprot_Protein_Name   \n",
       "0  cytoplasmic protein, YtiD                  Protein YtiD  \\\n",
       "1  cytoplasmic protein, YtiC                  Protein YtiC   \n",
       "2  cytoplasmic protein, YtiA  Uncharacterized protein YtiA   \n",
       "3  cytoplasmic protein, YpaA  Uncharacterized protein YpaA   \n",
       "4  cytoplasmic protein, YnfQ                  Protein YnfQ   \n",
       "\n",
       "  STEPdb_Sub-cellular_Location_Full Name   \n",
       "0                            Cytoplasmic  \\\n",
       "1                            Cytoplasmic   \n",
       "2                            Cytoplasmic   \n",
       "3                            Cytoplasmic   \n",
       "4                            Cytoplasmic   \n",
       "\n",
       "  STEPdb_Sub-cellular_Location_Letter Code sub-cellular_topology_group   \n",
       "0                                        A                 Cytoplasmic  \\\n",
       "1                                        A                 Cytoplasmic   \n",
       "2                                        A                 Cytoplasmic   \n",
       "3                                        A                 Cytoplasmic   \n",
       "4                                        A                 Cytoplasmic   \n",
       "\n",
       "  SRP_interactome_Schibich2016 TF_interactome_Oh2011  ...   \n",
       "0                          NaN                   NaN  ...  \\\n",
       "1                          NaN                   NaN  ...   \n",
       "2                          NaN                   NaN  ...   \n",
       "3                          NaN                   NaN  ...   \n",
       "4                          NaN                   NaN  ...   \n",
       "\n",
       "  TF_client_Martinez2009 TF_client_Butland2005 TF_client_Arifuzzaman2006   \n",
       "0                    NaN                   NaN                       NaN  \\\n",
       "1                    NaN                   NaN                       NaN   \n",
       "2                    NaN                   NaN                       NaN   \n",
       "3                    NaN                   NaN                       NaN   \n",
       "4                    NaN                   NaN                       NaN   \n",
       "\n",
       "  TF_client_Niwa2009 GroEL_client_Kerner2005 GroEL_client_Arifuzzaman2006   \n",
       "0                NaN                     NaN                          NaN  \\\n",
       "1                NaN                     NaN                          NaN   \n",
       "2                NaN                     NaN                          NaN   \n",
       "3                NaN                     NaN                          NaN   \n",
       "4                NaN                     NaN                          NaN   \n",
       "\n",
       "  GroEL_client_Chapman2006 GroEL_client_Niwa2009 SecB_client_Stepdb1.0   \n",
       "0                      NaN                   NaN                   NaN  \\\n",
       "1                      NaN                   NaN                   NaN   \n",
       "2                      NaN                   NaN                   NaN   \n",
       "3                      NaN                   NaN                   NaN   \n",
       "4                      NaN                   NaN                   NaN   \n",
       "\n",
       "  SecB_client_additional_data  \n",
       "0                         NaN  \n",
       "1                         NaN  \n",
       "2                         NaN  \n",
       "3                         NaN  \n",
       "4                         NaN  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# tratamento do xlsx chaperone interactors\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"datasets/MG1655_Chaperone_Interactors.csv\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
